node/libs/protobuf_build/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+//! buf breaking /tmp/sum.binpb --against /tmp/sum.binpb
+#![allow(clippy::print_stdout)]
+use anyhow::Context as _;
+pub use once_cell::sync::Lazy;
+// Imports accessed from the generated code.
+pub use prost;
+use prost::Message as _;
+pub use prost_reflect;
+use std::{fs, path::Path, sync::Mutex};
+pub use syntax::*;
+
+mod canonical;
+mod ident;
+mod syntax;
+
+/// Traversed all the files in a directory recursively.
+fn traverse_files(
+    path: &Path,
+    f: &mut impl FnMut(&Path) -> anyhow::Result<()>,
+) -> anyhow::Result<()> {
+    if !path.is_dir() {
+        f(path).with_context(|| path.to_str().unwrap().to_string())?;
+        return Ok(());
+    }
+    for entry in fs::read_dir(path)? {
+        traverse_files(&entry?.path(), f)?;
+    }
+    Ok(())
+}
+
+/// Protobuf descriptor + info about the mapping to rust code.
+pub struct Descriptor {
+    /// Root proto package that all proto files in this descriptor belong to.
+    /// Rust types have been generated relative to this root.
+    proto_root: ProtoName,
+    /// Raw descriptor proto.
+    descriptor_proto: prost_types::FileDescriptorSet,
+    /// Direct dependencies of this descriptor.
+    dependencies: Vec<&'static Descriptor>,
+}
+
+impl Descriptor {
+    /// Constructs a Descriptor.
+    pub fn new(
+        proto_root: ProtoName,
+        dependencies: Vec<&'static Descriptor>,
+        descriptor_bytes: &impl AsRef<[u8]>,
+    ) -> Self {
+        Descriptor {
+            proto_root,
+            dependencies,
+            descriptor_proto: prost_types::FileDescriptorSet::decode(descriptor_bytes.as_ref())
+                .unwrap(),
+        }
+    }
+
+    /// Loads the descriptor to the pool, if not already loaded.
+    pub fn load(&self, pool: &mut prost_reflect::DescriptorPool) -> anyhow::Result<()> {
+        if self
+            .descriptor_proto
+            .file
+            .iter()
+            .all(|f| pool.get_file_by_name(f.name()).is_some())
+        {
+            return Ok(());
+        }
+        for d in &self.dependencies {
+            d.load(pool)?;
+        }
+        pool.add_file_descriptor_set(self.descriptor_proto.clone())?;
+        Ok(())
+    }
+
+    /// Loads the descriptor to the global pool and returns a copy of the global pool.
+    pub fn load_global(&self) -> prost_reflect::DescriptorPool {
+        /// Global descriptor pool.
+        static POOL: Lazy<Mutex<prost_reflect::DescriptorPool>> = Lazy::new(Mutex::default);
+        let pool = &mut POOL.lock().unwrap();
+        self.load(pool).unwrap();
+        pool.clone()
+    }
+}
+
+/// Code generation config. Use it in build scripts.
+pub struct Config {
+    /// Input directory relative to $CARGO_MANIFEST_DIR with the proto files to be compiled.
+    pub input_root: InputPath,
+    /// Implicit prefix that should be prepended to proto paths of the proto files in the input directory.
+    pub proto_root: ProtoPath,
```

Comment:
Bikeshedding: IMO, here we should adhere to the "explicit is better than implicit" principle and require the Protobuf files always have the full directory structure. (E.g., `schema` files would be put in `proto/zksync/schema/` directory instead of `proto/`.)

Reply:
discussed offline, we will stay with implicit paths for now. If it turns out to be confusing to the users we can expand the paths later.